### ISSUE TITLE
siteUsers.getByLoginName and getByEmail throwing error when email has single quote 

### ISSUE THREAD

**Issue by hermilanastacio** [pnp/pnpjs#2889]
### What version of PnPjs library you are using

3.x

### Minor Version Number

18.0

### Target environment

All

### Additional environment details

Sharepoint Webpart/Authentication

### Question/Request

import { sp } from "@pnp/sp";

public async componentDidMount() {
await sp.web.siteUsers.getByLoginName("i:0#.f|membership|camp.o'[donnell@contoso.com.au](mailto:donnell@contoso.com.au)").get()
}

OR EVEN

public async componentDidMount() {
await sp.web.siteUsers.getByLoginName("i:0#.f|membership|camp.o'[donnell@contoso.com.au](mailto:donnell@contoso.com.au)")()
}

I'm getting 
"The expression "web/siteusers(@[vdriscoll@contoso.com.au](mailto:vdriscoll@contoso.com.au)')" is not valid."

Ps. I tried it in v2 and v3 but same error

Reference: https://github.com/pnp/pnpjs/issues/2874

**Comment by patrick-rodgers** [comment]
@hermilanastacio - if you encode the value before passing it - does that resolve your issue? I.e. you can replace the ' with '' (two single quotes). We'll have to trace through and cover all the various code paths to understand how best to address this.

**Comment by hermilanastacio** [comment]
@patrick-rodgers tried to encode it and also replacing ' with '' it but it throws error 400 (Bad Request)

![image](https://github.com/pnp/pnpjs/assets/12974206/c38c92b3-9dcb-4493-8c99-72966a30f323)

**Comment by bcameron1231** [comment]
I am unable to reproduce the issue using double single quotes. 
```
//works
await sp.web.siteUsers.getByEmail("camp.o''donnell@contoso.com.au")();
//works
await sp.web.siteUsers.getByLoginName("i:0#.f|membership|camp.o''donnell@contoso.com.au")()
```

**The error you just posted is coming from _api/web/ensureuser** Are you using sp.web.ensureUser() somewhere?

**Comment by hermilanastacio** [comment]
Here's the actual error message that I'm getting.
![image](https://github.com/pnp/pnpjs/assets/12974206/9125209d-c644-4b1b-b041-127873f46d25)

Endpoint -> https://{tenant}/sites/workplace/_api/web/siteusers(@vdriscoll%40contoso.com.au')?@v=%27i%3A0%23.f%7Cmembership%7Ccamp.o%27

**Comment by hermilanastacio** [comment]
Seems like in the url **https://{tenant}/sites/workplace/_api/web/siteusers(@v)?@v=%27i%3A0%23.f%7Cmembership%7Ccamp.o%27donnel%40contoso.com.au%27**, the value of @v= is the encoded login name **'i:0#.f|membership|camp.o'donnel@contoso.com.au'** with open and close single quote, and it's getting conflict with single quote in the email.

**Comment by hermilanastacio** [comment]
@patrick-rodgers yes, I tried it and it return an error **The query string "loginName" is missing or invalid.**

**Comment by hermilanastacio** [comment]
> I am unable to reproduce the issue using double single quotes.
> 
> ```
> //works
> await sp.web.siteUsers.getByEmail("camp.o''donnell@contoso.com.au")();
> //works
> await sp.web.siteUsers.getByLoginName("i:0#.f|membership|camp.o''donnell@contoso.com.au")()
> ```
> 
> However, **The error you just posted is coming from _api/web/ensureuser** not from these methods. Are you using sp.web.ensureUser() somewhere?

Yes I am, but that doesn't throw error on my end

**Comment by bcameron1231** [comment]
> Yes I am, but that doesn't throw error on my end

Based on the photo you sent before it was throwing an error. 

Looking back at your Postman photo, that is erroring because you're still only using a single quote, not a double single quote, thus you're getting the `queryString loginName is invalid` error.

As for the getByLoginName() in your latest photo (you're using V2), Yes I can confirm this doesn't work in V2. It will work in V3 when using double single quotes ( have tested this), and we recommend using V3. 

**To recap on my testing**
V2 - getByEmail() with double single quote - works
V2 - getByLoginName() - will not work as implemented
V3 - getByEmail() - with double single quote - works
V3 - getByLogin() - with double single quote - works.

Recommendation to try again. Use V3, and use single double quotes. I have tested this and it works.

**Comment by hermilanastacio** [comment]
@bcameron1231 , I used v3 and I tried these different text formats

![image](https://github.com/pnp/pnpjs/assets/12974206/0308e2b1-0a1e-417d-97ac-4ecc0afaab93)

but still, no luck. Could you please share how you do it?

**Comment by bcameron1231** [comment]
Can you confirm whether #3 is a double quote, or two single quotes? It's hard to tell from the pictures.

**Comment by bcameron1231** [comment]
Can you confirm what you're using? It looks like you're using double quotes. For example #3 and #4.

You should use two single quotes in a row.

**Comment by hermilanastacio** [comment]
@bcameron1231 
#1 I used single quote and just added (/) as escape character in email
#2 I used double quote for start and end of string
#3 I used double single quote and replace the single quote with double quote instead
#4 I tried string literal, I used backtick (`) character and replace the single quote with double quote instead
#5 I just tried string literal, I used backtick (`) character

**Comment by bcameron1231** [comment]
So #4 to confirm should be backtick, and the single quote between the name should be replaced with two single quotes, not a double quote.

**Comment by hermilanastacio** [comment]
@bcameron1231 tried it and I got same error on item #3

**Comment by bcameron1231** [comment]
I think your issue might still be coming from ensureUser and not getByLoginName. For clarity, ensureUser works without having to encode or change the loginname. getByEmail and getByLoginName however need to be modified as we've talked about. Here is y working code.

```
//no change to login name. Single quote, wrap login by double quotes
const ensureUser = await sp.web.ensureUser("i:0#.f|membership|tim.o'dare@cameroncode.onmicrosoft.com");
// Two single quotes between name, wrapped in double quotee
const user = await sp.web.siteUsers.getByLoginName("i:0#.f|membership|tim.o''dare@cameroncode.onmicrosoft.com")();
```
![image](https://github.com/pnp/pnpjs/assets/7944457/e52c984f-192e-463c-ae01-357dbfd80290)

![image](https://github.com/pnp/pnpjs/assets/7944457/5e05f4d7-4b71-4cf2-8efa-1a1d558f011b)

![image](https://github.com/pnp/pnpjs/assets/7944457/16d7cae2-932b-41a9-9adf-9c678e171279)

**Comment by hermilanastacio** [comment]
@bcameron1231 this code of yours works on v3

![image](https://github.com/pnp/pnpjs/assets/12974206/77ca6d92-7d58-484f-b585-24956fb68ae4)

need to change the single quote (') to double single quote ('') instead. Though my add to list function is not being pushed through after using v3 for some reason.

Anyways, now, I was able to get the right id of user with single quote in their email. Thank you so much for your help!

**Comment by bcameron1231** [comment]
Glad we were able to resolve it!
